### PR TITLE
Core/Players: Implement basics for the honor system

### LINKED ISSUE #18480: 7.x Reward PvP System
**Description:**
Does not the PvP rewards system work?

**Current behaviour:** (Tell us what happens.)

**Expected behaviour:** (Tell us what should happen instead.)

**Steps to reproduce the problem:**

Go to Battleground 

**Branch(es):** Master

**TC rev. hash/commit:**  Trinity Core Re Unknown

**TDB version:**  7.0.3

**Operating system:**   Windows 7 - 64bits

### PR DESCRIPTION
**Pull request by Golrag**
**Changes proposed:**
Implementing the following items:
- [x] Honor Leveling
- [x] Prestige
- [x] Rewards when honor leveling
- [x] CMSG_PVP_PRESTIGE_RANK_UP
- [x] Prestige and Honor Level Achievements
- [x] SPELL_EFFECT_GIVE_HONOR
- [x] Rested honor (saving in DB & consuming)
- [x] Prestige level 2 (Prestige level 1, honorlevel 50) reward is not working. Hackfixed.

**Target branch(es):** 3.3.5/master
- [x] master

**Issues addressed:** Closes #18480

**Tests performed:** (Does it build, tested in-game, etc.)
- Tested in game
- Builds

**Known issues and TODO list:**

Any feedback (and help) is welcome :)

### CODE REVIEW

File: src/server/game/DataStores/DB2Stores.h [hunk 1]
```diff
@@ -276,6 +280,7 @@ class TC_GAME_API DB2Manager
     typedef std::tuple<int16, int8, int32> WMOAreaTableKey;
     typedef std::map<WMOAreaTableKey, WMOAreaTableEntry const*> WMOAreaTableLookupContainer;
     typedef std::unordered_map<uint32, WorldMapAreaEntry const*> WorldMapAreaByAreaIDContainer;
+    typedef std::tuple<uint32, uint32> PvpRewardKey;
```
Comment (Shauren) [inline src/server/game/DataStores/DB2Stores.h]: you could have used std::pair for this and just went with regular unordered_map - we have a pair hasher in Common.h

Comment (Golrag) [inline src/server/game/DataStores/DB2Stores.h]: Thanks, updating :)

**Comment by Pixel-Pirate** [comment]
Good job

**Comment by Golrag** [comment]
Prestige level 1, honorlevel 50 rewards http://www.wowhead.com/item=138996/high-warlords-medal-of-valor for horde and http://www.wowhead.com/item=138992/grand-marshals-medal-of-valor for alliance

Sadly, I failed to find a link between RewardPackEntry 3 with unknown value 292 and those 2 items :(

File: src/server/game/DataStores/DB2Stores.cpp [hunk 2]
```diff
@@ -917,6 +925,12 @@ void DB2Manager::LoadStores(std::string const& dataPath, uint32 defaultLocale)
     for (WorldMapAreaEntry const* worldMapArea : sWorldMapAreaStore)
         _worldMapAreaByAreaID[worldMapArea->AreaID] = worldMapArea;
 
+    for (RewardPackXItemEntry const* rewardPackXItem : sRewardPackXItemStore)
+        _rewardPackXItems[rewardPackXItem->RewardPackID].push_back(rewardPackXItem);
+
+    for (PvpRewardEntry const* pvpReward : sPvpRewardStore)
+        _pvpRewardPack[std::pair<uint32, uint32>{ pvpReward->Prestige, pvpReward->HonorLevel }] = pvpReward->RewardPackID;
```
Comment (Warpten) [inline src/server/game/DataStores/DB2Stores.cpp]: `std::make_pair`

Comment (Golrag) [inline src/server/game/DataStores/DB2Stores.cpp]: Thanks, updating ! :)

Comment (Shauren) [inline src/server/game/DataStores/DB2Stores.cpp]: @Warpten please, you cant use make_pair with auto deduced type if it doesnt match the map key

Comment (Golrag) [inline src/server/game/DataStores/DB2Stores.cpp]: So ehm, should I leave it like this or revert my previous commit ?

Comment (Shauren) [inline src/server/game/DataStores/DB2Stores.cpp]: wait, does it compile? leave alone

Comment (Golrag) [inline src/server/game/DataStores/DB2Stores.cpp]: On local it does with VS2017 and it looks like travis is complaining about something else :p

**Comment by vanasis** [comment]
Try with [Achievement](http://www.wowhead.com/icon=236625/achievement-pvp-h-14#used-by-item) rewards, but I'm not sure

**Comment by Golrag** [comment]
I have tried to implement a hotfix in reward_pack which overrides the unknown value to 0 and put the correct data in hotfix_data for it but it didn't have any effect on the reward sadly so I don't know if it is related or not the the unknown value.. I might have made a mistake tho.

For achievement rewards, I don't know either. There is no achievement for being honor level 50 and prestige level 1

**Comment by Shauren** [comment]
What about this one? http://www.wowhead.com/achievement=10745

**Comment by Golrag** [comment]
This achievement is given when you finish the quest that's started with the item you get from honor leveling not sure if there is any link between those two

Actually you get the achievement by a spell cast on you after you finish the quest

**Comment by Golrag** [comment]
Things I tried so far:
- Tried to override the RewardPackID in PvpReward 31 (which is used for 1;50), client still shows the correct reward instead of the one I replaced it with. 

- Calling /script GetRewardPackItems(3); returns an empty array.

I could hardcode it in ```Player::SetHonorLevel```  but that's something I would like to avoid unless I'm 100% sure there is no other way to do it.

**Comment by DoctorKraft** [comment]
> Tried to override the RewardPackID in PvpReward 31
you have add `hotfix_data` for this entry with hash ?

**Comment by Aokromes** [comment]
/home/travis/build/TrinityCore/TrinityCore/src/server/game/Movement/MotionMaster.cpp:545:77: fatal error: 
      unused parameter 'spellEffectExtraData' [-Wunused-parameter]
  ...target /*= nullptr*/, Movement::SpellEffectExtraData const* spellEffectE...
                                                                 ^

**Comment by Shauren** [comment]
@Aokromes pls, that fails in master branch.

**Comment by Golrag** [comment]
@DoctorKraft yea, I should have put my SQL code tho

**Comment by DoctorKraft** [comment]
I have build your PR, i do not have look for rewards but i have found a problem. It's normal when you have Honor level 49 Prestige 1 and up 50 you are automatically prestige 2? Because on retail you have a box for choose to prestige or keep your level 50. I can help you for pvptalents if you need

**Comment by Golrag** [comment]
They changed it to automatically prestige when possible when they removed the 'reset pvp talents' feature upon prestiging, because there is really no downside to prestiging now.

For pvp talents, I decided that unlocking pvp talents, learning them, activating them in pvp only situations would be in another PR.

**Comment by DoctorKraft** [comment]
Oh ok, so no level 50 now ?
I have already much work for pvptalents maybe this can help you

**Comment by Golrag** [comment]
What do you mean by "no level 50 now?" ?

Right now when you level up to 49, you can get to level 50 but you automatically prestige setting your level back to 1. Unless you're max prestige level then you will stay 50. And no prestige button will appear.

I could make this a config option if there is any need for it

**Comment by DoctorKraft** [comment]
Currently when i reach level 49 and i give 1250 honor for up level 50, i'm automatically prestige 2. I don't have level 50 with 1250 honor for get prestige 2

**Comment by Golrag** [comment]
You don't need to gain honor once you're level 50 to prestige, client shows a full bar even if you have 0 honor after leveling to 50

**Comment by DoctorKraft** [comment]
Oops, ok. I think the problem with reward prestige 1 level 50 is faction (others stuffs with faction give a box) so maybe create a script called or update `achievement_rewards` when a player reach http://www.wowhead.com/achievement=10992/here-we-go-again ?

**Comment by Golrag** [comment]
While the effect would be the same now, it would not have been the case when you reached prestige level 1, honor level 50 back when you would not automatically prestige. Or when prestige level 2 is blocked for some reason.
I tried to remember if I received a mail with that item but it's been a while and found no entry 292 in MailTemplates.db2

**Comment by DoctorKraft** [comment]
Yep it's an instant reward item, look this video https://www.youtube.com/watch?v=QX3XoFpwKvo at right below gold you can look box "You winned" and autocly add quest (but this it's a flag in item)

**Comment by Shauren** [comment]
What about rested honor?

**Comment by Golrag** [comment]
Is that a thing ? :| I should look it up

**Comment by Shauren** [comment]
Oh yes it is, thats why PLAYER_FIELD_REST_INFO has bigger size in legion

**Comment by Golrag** [comment]
I reworked it a bit, added rested honor but I have no idea at which rate it should go up, any information on this ? All I have is https://twitter.com/holinka/status/779521793727995904 but using the same rate for normal xp for honor xp is way to fast

**Comment by Shauren** [comment]
No clue.

**Comment by Golrag** [comment]
Okay I think it's good now for the rested, could use some feedback (and help :)).
I moved it a bit out of player file, I can put it back if requested

File: src/server/game/Entities/Player/RestMgr.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+    float rest_bonus_max = float(_player->GetUInt32Value(next_level_xp_field)) * 1.5f / 2;
+
+    if (restBonus > rest_bonus_max)
+        RestBonus[restType] = rest_bonus_max;
+    else
+        RestBonus[restType] = restBonus;
+
+    // update data for client
+    if (_player->GetsRecruitAFriendBonus(true) && (_player->GetSession()->IsARecruiter() || _player->GetSession()->GetRecruiterId() != 0))
```
Comment (Shauren) [inline src/server/game/Entities/Player/RestMgr.cpp]: honor is not affected by RaF

**Comment by Eredar** [comment]
I guess "rested" honor is a 200% bar that you get every week. It appears not in the rest zones, but with weekly reset, and same amount every week, something around 3.5 honorlevels. I'm not too much in pvp but that's what I saw, maybe I'm wrong. Not 100% sure.

**Comment by wonder2011** [comment]
What about this PR ?, I thing is very important and need help and attention from the best devs of Trinity. I thing so !

**Comment by Golrag** [comment]
I would not like it if this PR gets merged without (hack) fixing the missing reward.
I'm still looking if that unknown value has something to do with it.

**Comment by DoctorKraft** [comment]
yeah need unknown value from RewardPack (last column) a lot of entry contains a value. (2 > 327)
Maybe it's related with GarrMissionReward deleted since Legion ? (no rewards also for this)

**Comment by Golrag** [comment]
I have thought about it and it seems that the last column of GarrMission.db2 is RewardPackID. Could be wrong tho, haven't looked much at it.

File: src/server/game/DataStores/DB2Structure.h [hunk 4]
```diff
@@ -2042,6 +2061,25 @@ struct RandPropPointsEntry
     uint32 UncommonPropertiesPoints[5];
 };
 
+struct RewardPackEntry
+{
+    uint32 ID;
+    uint32 Money;
+    float ArtifactPowerMultiplier;
```
Comment (Shauren) [inline src/server/game/DataStores/DB2Structure.h]: for consistency with quests I would like these fields to be called
```
        float  ArtifactXPMultiplier;
        uint32 ArtifactXPDifficulty;
        uint32 ArtifactCategoryID;
```

File: src/server/game/DataStores/DB2Structure.h [hunk 5]
```diff
@@ -2042,6 +2061,25 @@ struct RandPropPointsEntry
...
+    uint8 ArtifactQuestXPIndex;
+    uint8 ArtifactCategoryID;
+    uint32 TitleID;
+    uint32 Unk;
```
Comment (Shauren) [inline src/server/game/DataStores/DB2Structure.h]: This field is not used in the client, checked as far as the first legion beta build

**Comment by DoctorKraft** [comment]
Maybe you can find information here https://github.com/justMaku/DecompiledWoWLegionCompanion/

File: src/server/game/DataStores/DB2Structure.h (hunk 4 above)
Comment (Golrag) [inline src/server/game/DataStores/DB2Structure.h]: Wouldn't this mean that I have to change the metadata ? From ```ifbbii``` to ```ifiiii``` ?

File: src/server/game/DataStores/DB2Structure.h [hunk 6]
```diff
@@ -2065,11 +2065,11 @@ struct RewardPackEntry
 {
     uint32 ID;
     uint32 Money;
-    float ArtifactPowerMultiplier;
-    uint8 ArtifactQuestXPIndex;
-    uint8 ArtifactCategoryID;
+    float ArtifactXPMultiplier;
+    uint32 ArtifactXPDifficulty;
```
Comment (Shauren) [inline src/server/game/DataStores/DB2Structure.h]: but you shouldnt have changed the types, only names (i just copypasted these lines without thinking from quest code)

**Comment by Golrag** [comment]
Tested on retail for rested honor, and it seems like max level players can't get it, so I wonder why it's even there. If someone could tell me what happens when you reach level 110 while having rested honor, please share it :)

**Comment by Aokromes** [comment]
Conflicting files
src/server/game/Spells/SpellEffects.cpp

**Comment by dikkaa** [comment]
hi @Golrag 

We can't choose the talent ??
i just test the code and i can't choose the talent but everything else is good :)
Thank youu!

**Comment by wonder2011** [comment]
Merge ?

**Comment by Golrag** [comment]
@dikkaa honor talents etc will, eventually, be done in another PR.

**Comment by dikkaa** [comment]
@Golrag Alright thank you for the informations !

**Comment by wonder2011** [comment]
What about this pr?, Merge :P ??

**Comment by Golrag** [comment]
I would like to see how this PR works with 7.2 before it gets merged.
Blue post:

>As mentioned earlier in the thread, the amount of Honor required to level from Honor level 1-50 during Prestige 0 was cut in half in 7.2. Previous to 7.2, the amount of Honor required to level through Honor levels 1-10 was a very small number, so that characters would earn at least 1 Honor Talent in each tier very quickly.
Now that you no longer lose access to Honor Talents when you Prestige, the amount of Honor required for each level AFTER reaching Prestige 1 has been smoothed out. The total amount of Honor required for each Prestige level after Prestige 0 has been slightly reduced, not increased.

**Comment by Shauren** [comment]
the honor xp gametable got more columns

**Comment by Golrag** [comment]
Yea it got columns for every prestige level (0 - 16). Shouldn't be that hard to update.

**Comment by Golrag** [comment]
Changed ``` float Total;``` to ```float Prestige[16];``` in GameTables.h, but I would like to be able to increase the maximum prestige level without having to recompile, any ideas ?

**Comment by Golrag** [comment]
I might have misunderstand the last field in PrestigeLevelInfo.db2. Prestige level 5 - 8 and 14 - 19 got value 1, while 5 - 8 are listed.

**Comment by Golrag** [comment]
This PR should be ready :)

**Comment by Eredar** [comment]
In 7.0-7.1.5 rested honor was given at weekly reset. Something around 1-2 levels with 200% honorgain. Honor bar has slightly more bright color and separator with "rested" tooltip when you have it. Maybe it was changed in 7.2.

**Comment by Golrag** [comment]
Are you sure ? I logged off on one of my max level characters in rested area for two weeks and got 0 rested honor & no orange honor bar. While my level 100 character did the same, it had rested honor (confirmed in sniffs).

**Comment by Eredar** [comment]
"While my level 100 character did the same, it had rested honor (confirmed in sniffs)." Not sure about current state, but pre-7.2.0 honor system was 100% locked till max level (110). Low level characters recieved pvp badges and chests with items for winning games (not for every). Also its pretty hard to notice that "separator", I mentioned before. Since it appear only in the end of the "rested" bar. So you actually have to get 1 or 2 honorlevels to make it appear. That's what I remember about it in 7.1.5. Unsubbed untill ToS, so cant say for sure.

**Comment by Golrag** [comment]
It is still locked till max level. I have 0 proof that rested honor is a thing when honor leveling is enabled (lvl 110).
That's why I was kinda confused at first when Shauren mentioned rested honor. Not because I don't pvp but because I just never saw it, there is no indicator in game for non-max level characters for this and you don't get it on max level characters.

I'm not sure if this is intended behaviour from blizzard but what I am guessing is that "rested honor" is extra xp you get when you kill someone in a battleground or something similar (honor to xp).

>it works exactly like rested xp
>https://twitter.com/holinka/status/779521793727995904

Just logged in on my character who's been offline for a couple of months now in rested area, still no rested honor gained. So no changes from 7.1.x to 7.2 for this.

**Comment by Shauren** [comment]
I have not seen the rested honor on anything except fresh 110 myself

**Comment by Golrag** [comment]
Did you gain any extra rested honor after that ? It's something I was wondering about in this comment: https://github.com/TrinityCore/TrinityCore/pull/19277#issuecomment-288661092

**Comment by Shauren** [comment]
It looks like its based on amount of honor you earn over a certain period (does not simply stack for being logged out)
Examples
Fresh 110 (reached this week) had exactly 1400 rested honor (thats slightly over 10 levels)
char2 - prestige 1, level 4 also has rested honor
char3 - no prestige, level 24 has similar amount of rested honor
char4 - prestige 1, level 32, has rested honor
---- chars above only do pvp world quests if it rewards marks of honor (so not everyday)
main char - prestige 10, level 2, regularly clears pvp world quests every day, never has any rested honor

**Comment by Golrag** [comment]
Did you do any pvp on your fresh 110 before getting 110 ?

Still really confused at how it should work..
This is just speculating but...
Could it be that there is a difference in gaining rested honor before 110 and when 110 ?

level < 110
- You get rested honor while in inns etc. (may be consumed while doing battlegrounds)

level >= 110

- You need to have unlocked world quests. (my second 110 character is prestige level 1, and some honor levels, didn't play it in a couple of months and was dc'd in a rested area, but no world quests unlocked).
- On weekly reset (based on what Eredar said), check how much honor you could have made with (world quests * [some number]) =  x.
- Compare x with how much honor you gained.
- Do some math and you get your rested honor ?

**Comment by Shauren** [comment]
I don't know, but it will be good enough if you change this PR to just remove adding rested honor (keep consuming/saving to db)

**Comment by Golrag** [comment]
Done

File: src/server/game/DataStores/DB2Stores.cpp [hunk 7]
```diff
@@ -345,6 +351,8 @@ namespace
     ToyItemIdsContainer _toys;
     WMOAreaTableLookupContainer _wmoAreaTableLookup;
     WorldMapAreaByAreaIDContainer _worldMapAreaByAreaID;
+    PvpRewardsContainer _pvpRewardPack;
```
Comment (Shauren) [inline src/server/game/DataStores/DB2Stores.cpp]: sort, all of it is sorted

File: src/server/game/DataStores/DB2Stores.cpp [hunk 8]
```diff
@@ -1024,6 +1036,12 @@ void DB2Manager::LoadStores(std::string const& dataPath, uint32 defaultLocale)
     for (WorldMapAreaEntry const* worldMapArea : sWorldMapAreaStore)
         _worldMapAreaByAreaID[worldMapArea->AreaID] = worldMapArea;
 
+    for (RewardPackXItemEntry const* rewardPackXItem : sRewardPackXItemStore)
```
Comment (Shauren) [inline src/server/game/DataStores/DB2Stores.cpp]: sort

File: src/server/game/DataStores/DB2Stores.h [hunk 9]
```diff
@@ -322,6 +326,9 @@ class TC_GAME_API DB2Manager
     void Zone2MapCoordinates(uint32 areaId, float& x, float& y) const;
     void Map2ZoneCoordinates(uint32 areaId, float& x, float& y) const;
     static void DeterminaAlternateMapPosition(uint32 mapId, float x, float y, float z, uint32* newMapId = nullptr, DBCPosition2D* newPos = nullptr);
+    std::vector<RewardPackXItemEntry const*> const* GetRewardPackItemsByRewardID(uint32 rewardPackID) const;
```
Comment (Shauren) [inline src/server/game/DataStores/DB2Stores.h]: pls, keep the sorting (ordered by related db2 name)

Comment (Golrag) [inline src/server/game/DataStores/DB2Stores.h]: Should I also sort them in DB2Stores.cpp ?

Comment (Shauren) [inline src/server/game/DataStores/DB2Stores.h]: ofc thats a yes

**Comment by Shauren** [comment]
I found criteria tree that have a completion requirement being prestige 1 honor level 50
Tree ids: 51140, 51156 (criteria id 31773, modifier tree id 37759) - they are not attached to any achievement

File: src/server/game/DataStores/DB2Stores.cpp [hunk 10]
```diff
@@ -296,6 +300,8 @@ typedef std::unordered_set<uint32> ToyItemIdsContainer;
 typedef std::tuple<int16, int8, int32> WMOAreaTableKey;
 typedef std::map<WMOAreaTableKey, WMOAreaTableEntry const*> WMOAreaTableLookupContainer;
 typedef std::unordered_map<uint32, WorldMapAreaEntry const*> WorldMapAreaByAreaIDContainer;
+typedef std::unordered_map<std::pair<uint32 /*prestige level*/, uint32 /*honor level*/>, uint32> PvpRewardsContainer;
```
Comment (Shauren) [inline src/server/game/DataStores/DB2Stores.cpp]: missed sorting for this

File: src/server/game/Entities/Player/Player.h [hunk 11]
```diff
@@ -2730,6 +2715,9 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
         SceneMgr m_sceneMgr;
 
         std::unordered_map<ObjectGuid /*LootObject*/, ObjectGuid /*world object*/> m_AELootView;
+
+        void _InitHonorLevelOnLoadFromDB(uint32 /*honor*/, uint32 /*honorLevel*/, uint32 /*prestigeLevel*/);
+        RestMgr* _restMgr;
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.h]: just use std::unique_ptr for this

**Comment by Shauren** [comment]
Ok, these should be the last changes needed before merge

**Comment by Golrag** [comment]
>I found criteria tree that have a completion requirement being prestige 1 honor level 50

Happy to hear this :). Added a comment to the code for it.

File: src/server/game/Spells/SpellEffects.cpp [hunk 12]
```diff
@@ -5876,3 +5877,22 @@ void Spell::EffectGiveArtifactPowerNoBonus(SpellEffIndex /*effIndex*/)
...
+
+void Spell::EffectGiveHonor(SpellEffIndex /*effIndex*/)
+{
+    if (effectHandleMode != SPELL_EFFECT_HANDLE_HIT_TARGET)
+        return;
+
+    if (!unitTarget || unitTarget->GetTypeId() != TYPEID_PLAYER)
+        return;
+
+    int32 honor = effectInfo->CalcValue();
```
Comment (joschiwald) [inline src/server/game/Spells/SpellEffects.cpp]: that is `damage`

File: src/server/game/Spells/SpellEffects.cpp [hunk 13]
```diff
@@ -5876,3 +5877,22 @@ void Spell::EffectGiveArtifactPowerNoBonus(SpellEffIndex /*effIndex*/)
...
+
+    WorldPackets::Combat::PvPCredit packet;
+    packet.Honor = honor;
+    packet.OriginalHonor = honor;
+
+    Player* playerTarget = unitTarget->ToPlayer();
+    playerTarget->AddHonorXP(honor);
+    playerTarget->GetSession()->SendPacket(packet.Write());
```
Comment (joschiwald) [inline src/server/game/Spells/SpellEffects.cpp]: `player->SendDirectMessage(packet.Write());`

File: src/server/game/Spells/SpellEffects.cpp [hunk 14]
```diff
@@ -5876,3 +5877,20 @@ void Spell::EffectGiveArtifactPowerNoBonus(SpellEffIndex /*effIndex*/)
...
+{
+    if (effectHandleMode != SPELL_EFFECT_HANDLE_HIT_TARGET)
+        return;
+
+    if (!unitTarget || unitTarget->GetTypeId() != TYPEID_PLAYER)
+        return;
+
+    WorldPackets::Combat::PvPCredit packet;
+    packet.Honor = damage;
+    packet.OriginalHonor = damage;
```
Comment (joschiwald) [inline src/server/game/Spells/SpellEffects.cpp]: i don't think that OriginalHonor is the same like Honor
could be Honor the real honor gain?

Comment (Golrag) [inline src/server/game/Spells/SpellEffects.cpp]:
```
Player::RewardHonor
```

Uses the same value for both

Don't think I have a sniff from retail for it :/
Also they might have changed something for 7.2 as rank isn't showed to the client anymore. When you kill an enemy player, you just get the message that you gained honor (no player rank message given).

**Pull request closed without merging**